### DISCUSSION TITLE
Where is my local traffic?

### DISCUSSION THREAD

**Discussion by nirnachmani** [mitmproxy/mitmproxy#5723] [Q&A] [answered]
I'm quite lost and wondering if someone more experienced would be able to help. I am trying to monitor traffic between an android app that controls smart lights and a propriety hub that actually controls the lights. I have been successfully using mitmproxy to monitor traffic from a virtual android device running the app to reverse engineer the communication so to re-create it in python. Until recently I have assumed that the app is using the cloud/internet for all communications. However, I have now realized that it has been using the cloud because by using the virtual android device I created a separate LAN which prevented local communication between the app and the hub. When both the app and the hub are on the same LAN, the communication is local. I now want to monitor the local communication so I can implement it in my script. 

For that, I rooted an old android phone, installed the mitmproxy CA certificate as a system certificate and installed the app so now this phone communicates with the hub locally. However, when looking at traffic from that phone on mitmproxy, I can only see traffic that is directed to the internet. I have tried both regular mode and transparent mode and I can see traffic (HTTP, HTTPS and TCP) from the app in both modes, but no local traffic directed to the hub is showing up. I found out which ports the app uses (by using lsof) and directed the port to port 8080 with the ip6tables command, but still no local traffic. I am sure there is local traffic I am not seeing because when I turn off access to the internet the app can still communicate with the hub to control the lights, but no traffic is showing on mitmproxy. 

I tried running mitmproxy with udp_hosts but I get an error (mitmproxy: error: unrecognized arguments: --udp-hosts). 

What am I missing? is the app using UDP (although lsof is not showing UDP ports)? is the local traffic somehow bypasses mitmproxy (I disabled ICMP redirects)? or am I just not setting things correctly to capture local traffic? I assumed that in transparent mode with default gateway on the phone set to the mitmproxy host, it should capture all traffic, including local. 

My mitmproxy is running on Ubuntu.  

Thank you.

**Comment by mhils** [comment, marked as answer]
Have you tried using WireGuard mode? :)

**Reply by nirnachmani** [reply]
Funny you say that. I just did and then saw your post. And the data is there! It's in a TCP connection, keeps updating. 

There are also some UDP broadcasts but I can't see their data - it says "[Errno 13] Permission denied". I'm also getting a TLS handshake error occasionally "Server TLS handshake failed. Certificate verify failed: self-signed certificate", It is originating from different ports. I'm not sure if this comes from the app I'm interested in or other apps. It may be why I can't see the UDP data. 

Thanks!

**Reply by mhils** [reply]
> it says "[Errno 13] Permission denied". 

Could you clarify where/how you see that error message? Does that also happen if you run mitmproxy as root/admin?

**Reply by nirnachmani** [reply]
![Screenshot from 2022-11-06 08-10-46](https://user-images.githubusercontent.com/83867579/200142326-ff46def9-3e83-4d0c-afbb-6c2c37fab9a9.png)

Only happens with UDP from that source (10.0.0.1:41580) 

It also happens when I run it is root. When I run it as root it kept coming with the following error message: 
```
Client TLS handshake failed. The client does not trust the proxy's certificate for www.pixie.app (OpenSSL Error([('SSL routines', '', 'sslv3 alert certificate unknown')]))
```
(www.pixie.app is the app I am reverse engineering) 

However, I don't get this message when not running as root. I get the following message instead:

```
error establishing server connection: [Errno 13] Permission denied
```
